Mathlib/Geometry/Manifold/InteriorBoundary.lean

```diff
@@ -162,43 +170,53 @@ variable {I}
 
 /-- The interior of `M × N` is the product of the interiors of `M` and `N`. -/
-lemma ModelWithCorners.interior_prod :
+lemma interior_prod :
     (I.prod J).interior (M × N) = (I.interior M) ×ˢ (J.interior N) := by
   ext p
   have aux : (interior (range ↑I)) ×ˢ (interior (range J)) = interior (range (I.prod J)) := by
     rw [← interior_prod_eq, ← Set.range_prod_map, modelWithCorners_prod_coe]
   constructor <;> intro hp
   · replace hp : (I.prod J).IsInteriorPoint p := hp
-    rw [ModelWithCorners.IsInteriorPoint, ← aux] at hp
+    rw [IsInteriorPoint, ← aux] at hp
     exact hp
-  · obtain ⟨h₁, h₂⟩ := Set.mem_prod.mp hp
-    rw [ModelWithCorners.interior] at h₁ h₂
-    show (I.prod J).IsInteriorPoint p
-    rw [ModelWithCorners.IsInteriorPoint, ← aux, mem_prod]
-    exact ⟨h₁, h₂⟩
+  · show (I.prod J).IsInteriorPoint p
+    rw [IsInteriorPoint, ← aux, mem_prod]
+    obtain h := Set.mem_prod.mp hp
+    rw [ModelWithCorners.interior] at h
+    exact h
 
 /-- The boundary of `M × N` is `∂M × N ∪ (M × ∂N)`. -/
-lemma ModelWithCorners.boundary_prod :
+lemma boundary_prod :
     (I.prod J).boundary (M × N) = Set.prod univ (J.boundary N) ∪ Set.prod (I.boundary M) univ := by
   let h := calc (I.prod J).boundary (M × N)
     _ = ((I.prod J).interior (M × N))ᶜ := (I.prod J).boundary_eq_complement_interior
-    _ = ((I.interior M) ×ˢ (J.interior N))ᶜ := by rw [ModelWithCorners.interior_prod]
+    _ = ((I.interior M) ×ˢ (J.interior N))ᶜ := by rw [interior_prod]
     _ = (I.interior M)ᶜ ×ˢ univ ∪ univ ×ˢ (J.interior N)ᶜ := by rw [compl_prod_eq_union]
   rw [h, I.boundary_eq_complement_interior, J.boundary_eq_complement_interior, union_comm]
   rfl
 
 /-- If `M` is boundaryless, `∂(M×N) = M × ∂N`. -/
-lemma boundary_of_boundaryless_left [I.Boundaryless] :
+lemma boundary_of_boundaryless_left [BoundarylessManifold I M] :
     (I.prod J).boundary (M × N) = Set.prod (univ : Set M) (J.boundary N) := by
-  rw [ModelWithCorners.boundary_prod, ModelWithCorners.Boundaryless.boundary_eq_empty I]
+  rw [boundary_prod, Boundaryless.boundary_eq_empty I]
   have : Set.prod (∅ : Set M) (univ : Set N) = ∅ := Set.empty_prod
   rw [this, union_empty]
 
 /-- If `N` is boundaryless, `∂(M×N) = ∂M × N`. -/
-lemma boundary_of_boundaryless_right [J.Boundaryless] :
+lemma boundary_of_boundaryless_right [BoundarylessManifold J N] :
     (I.prod J).boundary (M × N) = Set.prod (I.boundary M) (univ : Set N) := by
-  rw [ModelWithCorners.boundary_prod, ModelWithCorners.Boundaryless.boundary_eq_empty J]
+  rw [boundary_prod, Boundaryless.boundary_eq_empty J]
   have : Set.prod (univ : Set M) (∅ : Set N) = ∅ := Set.prod_empty
   rw [this, empty_union]
 
+/-- The product of two boundaryless manifolds is boundaryless. -/
+instance BoundarylessManifold.prod [BoundarylessManifold I M] [BoundarylessManifold J N] :
+    BoundarylessManifold (I.prod J) (M × N) := by
+  apply Boundaryless.of_boundary_eq_empty
+  rw [boundary_of_boundaryless_left, Boundaryless.boundary_eq_empty]
```

Comment:
Does `simp [boundary_prod, Boundaryless.boundary_eq_empty]` work?

Reply:
It works to replace lines 2+ by the following
```
  simp [boundary_prod, Boundaryless.boundary_eq_empty]
  constructor
  · exact Set.prod_empty
  · exact Set.empty_prod
```
However, `simp` by itself does not apply these lemmas (even though they are marked simp). Supplying them explicitly doesn't work either. If you know a trick to get simp to use them, I'm all ears.

Reply:
Probably, this is the `Prod` vs `ModelProd` issue.

Reply:
Makes sense. I've left a comment.